### PR TITLE
[IMP] mail, *: res.users.settings improvements

### PR DESCRIPTION
**Pull request by oomsveta**
\* = im_livechat, test_discuss_full

Various improvements to user settings, including:

+ Introduce new js models `res.users.settings` and `res.users.settings.volumes`
to mirror the python models.
+ Avoid reformatting server data and simplify the way notifications are handled.
+ Clean dead code (unused convertData).
+ Reword and add documentation.

Enterprise: https://github.com/odoo/enterprise/pull/28291

### CODE REVIEW

File: addons/mail/static/src/models/messaging_initializer.js [hunk 1]
```diff
@@ -221,28 +221,28 @@ registerModel({
...
-         * @param {Object} [resUsersSettings.volume_settings]
+         * @param {Object} [resUsersSettings.volume_settings_ids]
          */
         _initResUsersSettings({
             id,
             is_discuss_sidebar_category_channel_open,
             is_discuss_sidebar_category_chat_open,
-            use_push_to_talk,
             push_to_talk_key,
+            use_push_to_talk,
```
Comment (alexkuhn) [inline addons/mail/static/src/models/messaging_initializer.js]: 1st commit: useless diff of reordering

File: addons/mail/static/src/models/messaging_notification_handler.js [hunk 2]
```diff
@@ -452,12 +451,10 @@ registerModel({
...
          * @param {Object} data
-         * @param {Object} [data.volumeSettings]
+         * @param {Object} [data.volume_settings_ids]
          */
-        async _handleNotificationVolumeSettingUpdate({ volumeSettings }) {
-            this.messaging && this.messaging.userSetting.update({
-                volumeSettings: volumeSettings,
-            });
+        async _handleNotificationVolumeSettingUpdate({ volume_settings_ids }) {
+            this.messaging.userSetting.update({ volume_settings_ids });
```
Comment (alexkuhn) [inline addons/mail/static/src/models/messaging_notification_handler.js]: 1st commit: useless syntax reformating

File: addons/mail/static/src/models/messaging_notification_handler.js [hunk 3]
```diff
@@ -388,7 +388,6 @@ registerModel({
          * @param {number} settings.voice_active_duration
          * @param {boolean} [settings.is_discuss_sidebar_category_channel_open]
          * @param {boolean} [settings.is_discuss_sidebar_category_chat_open]
-         * @param {Object} [payload.volume_settings]
```
Comment (alexkuhn) [inline addons/mail/static/src/models/messaging_notification_handler.js:391]: 1st commit: cleaning of docstring is unrelated to this commit, which is rename. Could be in another "cleaning" commit

File: addons/mail/static/src/models/discuss_sidebar_category.js [hunk 4]
```diff
@@ -403,7 +415,9 @@ registerModel({
         /**
          * Boolean that determines the last open state known by the server.
          */
-        isServerOpen: attr(),
+        isServerOpen: attr({
+            compute: '_computeIsServerOpen',
+        }),
```
Comment (alexkuhn) [inline addons/mail/static/src/models/discuss_sidebar_category.js:428]: 3rd commit: missing `default: false` here

File: addons/mail/static/src/models/discuss_sidebar_category.js [hunk 5]
```diff
@@ -145,6 +145,18 @@ registerModel({
...
+         * @returns {boolean}
+         */
+        _computeIsServerOpen() {
+            if (this.discussAsChannel) {
+                return this.messaging.userSetting.is_discuss_sidebar_category_channel_open;
+            }
+            if (this.discussAsChat) {
+                return this.messaging.userSetting.is_discuss_sidebar_category_chat_open;
+            }
+        },
```
Comment (alexkuhn) [inline addons/mail/static/src/models/discuss_sidebar_category.js:167]: 3rd commit: missing `return clear()` at the end

File: addons/mail/static/src/models/user_setting.js [hunk 6]
```diff
@@ -132,9 +131,9 @@ registerModel({
             await this.messaging.rtc.updateVoiceActivation();
         },
         async togglePushToTalk() {
-            this.update({ use_push_to_talk: !this.use_push_to_talk });
+            this.update({ isPushToTalkUsed: !this.isPushToTalkUsed });
             await this.messaging.rtc.updateVoiceActivation();
-            if (!this.messaging.isCurrentUserGuest) {
+            if (this.messaging.currentUser) {
```
Comment (alexkuhn) [inline addons/mail/static/src/models/user_setting.js:136]: 4th commit: this change from `!this.messaging.isCurrentUserGuest` to `this.messaging.currentUser` can be in its own commit

File: addons/mail/static/src/models/user_setting.js [hunk 7]
```diff
@@ -97,9 +96,9 @@ registerModel({
          * @param {event} ev
          */
         async setPushToTalkKey(ev) {
-            const push_to_talk_key = `${ev.shiftKey || ''}.${ev.ctrlKey || ev.metaKey || ''}.${ev.altKey || ''}.${ev.key}`;
-            this.update({ push_to_talk_key });
-            if (!this.messaging.isCurrentUserGuest) {
+            const pushToTalkKey = `${ev.shiftKey || ''}.${ev.ctrlKey || ev.metaKey || ''}.${ev.altKey || ''}.${ev.key}`;
+            this.update({ pushToTalkKey });
+            if (this.messaging.currentUser) {
```
Comment (alexkuhn) [inline addons/mail/static/src/models/user_setting.js:101]: same here

File: addons/mail/static/src/models/user_setting.js [hunk 8]
```diff
@@ -260,31 +285,35 @@ registerModel({
...
             isCausal: true,
         }),
+        globalSettingsTimeout: attr(),
         /**
-         * true if the user wants to use push to talk (over voice activation)
+         * If true, push-to-talk will be used over voice activation.
          */
-        use_push_to_talk: attr({
+        isPushToTalkUsed: attr({
+            compute: '_computeIsPushToTalkUsed',
```
Comment (alexkuhn) [inline addons/mail/static/src/models/user_setting.js:293]: Diff is hard to read because the old field name has been moved to new model, and here it's a rename and becomes related.

It would be much easier to split in smaller commit with small and very clear concern. For example, one commit should introduce the model (just the model, nothing else), then another commit to move the fields, then another commit to introduce compute fields from `UserSetting` to `res.users.settings`, etc.

This is maybe more diff overall, but it's significantly much easier and safer to review than current diff in a single commit.

File: addons/mail/static/src/models/res_users_settings.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        push_to_talk_key: attr(),
+        use_push_to_talk: attr(),
+        user_id: one('User', {
+            inverse: 'resUsersSettings',
+        }),
+        voice_active_duration: attr(),
+        /**
+         * States the volume chosen by the current user for each other user.
+         */
+        volume_settings_ids: many('VolumeSetting', {
```
Comment (alexkuhn) [inline addons/mail/static/src/models/res_users_settings.js:28]: I would expect this to be `res.users.settings.volumes`, a python model.

File: addons/mail/static/src/models/res_users_settings.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    fields: {
+        id: attr({
+            readonly: true,
+            required: true,
+        }),
+        is_discuss_sidebar_category_channel_open: attr(),
+        is_discuss_sidebar_category_chat_open: attr(),
+        push_to_talk_key: attr(),
+        use_push_to_talk: attr(),
+        user_id: one('User', {
```
Comment (alexkuhn) [inline addons/mail/static/src/models/res_users_settings.js:21]: I would expect this to be `res.users`, a python model.

**Review by alexkuhn — APPROVED** [review]
Overall LGTM

Can do small changes here and there, but I'd say keep it small and fast enough so that this PR is merged soon.

The part that introduces model `res.users.settings` and still does shortcut with relation to JS models rather than python models `res.users` and `res.users.settings.volume`, I'm good not doing it now, just because could slow down merging this PR, and it's really not mandatory.

File: addons/mail/models/res_users_settings.py [hunk 11]
```diff
@@ -32,11 +32,8 @@ def _find_or_create_for_user(self, user):
     def _res_users_settings_format(self):
         self.ensure_one()
         res = self._read_format(fnames=[name for name, field in self._fields.items() if name == 'id' or not field.automatic])[0]
-        res.pop('volume_settings_ids')
         volume_settings = self.volume_settings_ids._discuss_users_settings_volume_format()
-        res.update({
-            'volume_settings': [('insert', volume_settings)] if volume_settings else [],
-        })
+        res['volume_settings_ids'] = [('insert', volume_settings)] if volume_settings else []
```
Comment (seb-odoo) [inline addons/mail/models/res_users_settings.py:36]: if/else probably not necessary? insert empty array is fine

File: addons/mail/static/src/models/res_users_settings.js (hunk 10 above)
Comment (seb-odoo) [reply addons/mail/static/src/models/res_users_settings.js:27]: Let's not introduce all of those models now, it's probably going to be more trouble than it's worth.

File: addons/mail/static/src/models/res_users_settings.js (hunk 9 above)
Comment (seb-odoo) [reply addons/mail/static/src/models/res_users_settings.js:35]: For volume settings specifically though we should introduce it now, as it is directly in the scope of this PR.

File: addons/im_livechat/static/src/models/messaging_initializer.js [hunk 12]
```diff
@@ -8,17 +8,14 @@ import '@mail/models/messaging_initializer';
...
-     * @param {Object} resUsersSettings
-     * @param {boolean} resUsersSettings.is_discuss_sidebar_category_livechat_open
      */
-    _initResUsersSettings({ is_discuss_sidebar_category_livechat_open }) {
-        this._super(...arguments);
-        this.messaging.userSetting.update({ is_discuss_sidebar_category_livechat_open });
+    _initResUsersSettings(...args) {
         this.messaging.discuss.update({
             categoryLivechat: insertAndReplace({
                 serverStateKey: 'is_discuss_sidebar_category_livechat_open',
```
Comment (seb-odoo) [inline addons/im_livechat/static/src/models/messaging_initializer.js:15]: This should be a compute too, why isn't it removed in the commit doing the others?

File: addons/im_livechat/static/src/models/messaging_notification_handler.js [hunk 13]
```diff
@@ -5,19 +5,6 @@ import { patchRecordMethods } from '@mail/model/model_core';
...
-     * @param {object} settings
-     * @param {boolean} [settings.is_discuss_sidebar_category_livechat_open]
-    */
-    _handleNotificationResUsersSettings(settings) {
-        if ('is_discuss_sidebar_category_livechat_open' in settings) {
-            this.messaging.userSetting.update({
-                is_discuss_sidebar_category_livechat_open: settings.is_discuss_sidebar_category_livechat_open,
-            });
-        }
-        this._super(settings);
```
Comment (seb-odoo) [inline addons/im_livechat/static/src/models/messaging_notification_handler.js:19]: confusing commit split, this should be removed in the commit introducing the computes

File: addons/im_livechat/static/tests/helpers/mock_server.js [hunk 14]
```diff
@@ -252,4 +252,14 @@ patch(MockServer.prototype, 'im_livechat', {
...
+    /**
+     * @override
+     */
+    _mockResUsersSettings_ResUsersSettingsFormat(id) {
+        const [settings] = this.getRecords('res.users.settings', [['id', '=', id]]);
+        return {
+            ...this._super(...arguments),
+            is_discuss_sidebar_category_livechat_open: settings.is_discuss_sidebar_category_livechat_open,
+        };
+    },
```
Comment (seb-odoo) [inline addons/im_livechat/static/tests/helpers/mock_server.js:264]: there should be no override, and base code returning all fields instead

also this whole "use res.users.settings fields in JS code" commit is really badly split

makes no sense to have change in server or mock server in such commit

Comment (oomsveta) [reply addons/im_livechat/static/tests/helpers/mock_server.js:264]: > and base code returning all fields instead

Without formatting, we retrieve unwanted fields added by the MockServer (`__last_update`, `name`, `display_name`), and all the automatic fields that are normally filtered server side.

Implementing the same logic here would require to list all the fields to be removed, is that what I should do?

Comment (seb-odoo) [reply addons/im_livechat/static/tests/helpers/mock_server.js:264]: Yes I prefer to list the fields to remove in this case, same as in python in the mock server.

File: addons/mail/static/src/models/rtc_session.js [hunk 15]
```diff
@@ -74,15 +74,15 @@ registerModel({
...
+            }
             if (this.partner && this.partner.volumeSetting) {
                 this.partner.volumeSetting.update({ volume });
             }
             if (this.guest && this.guest.volumeSetting) {
                 this.guest.volumeSetting.update({ volume });
             }
-            if (this.messaging.isCurrentUserGuest) {
-                return;
-            }
```
Comment (seb-odoo) [inline addons/mail/static/src/models/rtc_session.js:85]: bad move (at least in this commit)

File: addons/mail/static/src/models/user.js [hunk 16]
```diff
@@ -195,9 +195,8 @@ registerModel({
         partner: one('Partner', {
             inverse: 'user',
         }),
-        /**
-         * Id of this user's res.users.settings record.
-         */
-        resUsersSettingsId: attr(),
+        resUsersSettings: one('res.users.settings', {
```
Comment (seb-odoo) [inline addons/mail/static/src/models/user.js:198]: should be `res_users_settings_id` like in python

File: addons/mail/static/src/models/rtc.js [hunk 17]
```diff
@@ -312,7 +312,7 @@ registerModel({
             if (this.disconnectAudioMonitor) {
                 this.disconnectAudioMonitor();
             }
-            if (this.messaging.userSetting.usePushToTalk || !this.channel || !this.audioTrack) {
+            if (this.messaging.userSetting.isPushToTalkUsed || !this.channel || !this.audioTrack) {
```
Comment (seb-odoo) [inline addons/mail/static/src/models/rtc.js:315]: this rename is adding a lot of noise for not much, and the new name is not really better

File: addons/mail/static/tests/helpers/mock_server.js [hunk 18]
```diff
@@ -2241,6 +2256,20 @@ patch(MockServer.prototype, 'mail', {
...
 
+    _mockResUsersSettingsVolumes_DiscussUsersSettingsVolumeFormat(ids) {
+        const volumeSettingsRecords = this.getRecords('res.users.settings.volumes', [['id', 'in', ids]]);
+        return volumeSettingsRecords.map(volumeSettingsRecord => {
+            const [relatedGuest] = this.getRecords('mail.guest', [['id', '=', volumeSettingsRecord.guest_id]]);
+            const [relatedPartner] = this.getRecords('res.partner', [['id', '=', volumeSettingsRecord.partner_id]]);
+            return {
+                guest: relatedGuest ? [['insert-and-replace', { id: relatedGuest.id, name: relatedGuest.name }]] : [['clear']],
+                id: volumeSettingsRecord.id,
+                partner: relatedPartner ? [['insert-and-replace', { id: relatedPartner.id, name: relatedPartner.name }]] : [['clear']],
```
Comment (seb-odoo) [inline addons/mail/static/tests/helpers/mock_server.js:2267]: guest_id, partner_id ?

File: addons/mail/static/src/models/partner.js [hunk 19]
```diff
@@ -436,8 +436,8 @@ registerModel({
         user: one('User', {
             inverse: 'partner',
         }),
-        volumeSetting: one('VolumeSetting', {
-            inverse: 'partner',
+        volumeSetting: one('res.users.settings.volumes', {
```
Comment (seb-odoo) [inline addons/mail/static/src/models/partner.js:439]: `res_user_settings_volumes_id`?

File: addons/mail/static/src/models/messaging_notification_handler.js [hunk 20]
```diff
@@ -452,12 +432,10 @@ registerModel({
...
          * @param {Object} data
-         * @param {Object} [data.volumeSettings]
+         * @param {Object} [data.volume_settings_ids]
          */
-        async _handleNotificationVolumeSettingUpdate({ volumeSettings }) {
-            this.messaging && this.messaging.userSetting.update({
-                volumeSettings: volumeSettings,
-            });
+        async _handleNotificationVolumeSettingUpdate({ volume_settings_ids }) {
+            this.messaging.currentUser.resUsersSettings.update({ volume_settings_ids });
```
Comment (seb-odoo) [inline addons/mail/static/src/models/messaging_notification_handler.js:438]: param just data, and update data?
Also this should ideally become just insert on `res.users.settings.volumes` with id provided to identify it

and notif type just `res.users.settings.volumes/insert`

File: addons/mail/static/src/models/messaging_initializer.js [hunk 21]
```diff
@@ -225,44 +219,10 @@ registerModel({
...
-                categoryChannel: insertAndReplace({
-                    isServerOpen: is_discuss_sidebar_category_channel_open,
-                    serverStateKey: 'is_discuss_sidebar_category_channel_open',
-                }),
-                categoryChat: insertAndReplace({
-                    isServerOpen: is_discuss_sidebar_category_chat_open,
-                    serverStateKey: 'is_discuss_sidebar_category_chat_open',
-                }),
-            });
+        _initResUsersSettings(resUsersSettingsData) {
```
Comment (seb-odoo) [inline addons/mail/static/src/models/messaging_initializer.js:224]: overkill as a separate method?

File: addons/mail/static/src/models/messaging_initializer.js [hunk 22]
```diff
@@ -225,44 +219,10 @@ registerModel({
...
+            this.messaging.currentUser.update({ resUsersSettings: insertAndReplace(resUsersSettingsData) });
```
Comment (seb-odoo) [inline addons/mail/static/src/models/messaging_initializer.js:225]: ideally should be insert global, link to user should be done through user_id field of setting

File: addons/mail/static/src/models/guest.js [hunk 23]
```diff
@@ -46,8 +46,8 @@ registerModel({
         rtcSessions: many('RtcSession', {
             inverse: 'guest',
         }),
-        volumeSetting: one('VolumeSetting', {
-            inverse: 'guest',
+        volumeSetting: one('res.users.settings.volumes', {
```
Comment (seb-odoo) [inline addons/mail/static/src/models/guest.js:49]: `res_users_settings_volumes_id`?

**Review by seb-odoo — APPROVED** [review]
@robodoo r+
